Persistent UserManagedCache hits NPE

Client version:
````
Implementation-Title: ehcache
Implementation-Version: 3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3
 e00709639
````

Settings:
````
Threads: 4
Elements: 100000
Configuration: heap(1) _disk(2GB)
Variation: [Cache level copier variation: [Key copier: LongCopier ,Value copier: CacheValueCopier], Cache level serializer variation: [Key serializer: LongSerializer ,Value serializer: CacheValueSerializer], TTL Expiry, Eviction Veto None]
````

In the scenario, puts done UserManagedCache configured with above settings and closed the cache, further when trying to reload the persisted cache data hits the below exception.

Please note that the same test worked fine sometime ago (don't have exact date/build history) 

````
16:34:38.921 [pool-6-thread-1] INFO  o.e.c.i.r.LoggingRobustResilienceStrategy - Ehcache key 71791 recovered from
org.ehcache.core.spi.store.StoreAccessException: java.lang.NullPointerException
	at org.ehcache.impl.internal.store.heap.OnHeapStore.getOrComputeIfAbsent(OnHeapStore.java:747) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.impl.internal.store.tiering.TieredStore.get(TieredStore.java:110) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.core.Ehcache.get(Ehcache.java:167) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.core.PersistentUserManagedEhcache.get(PersistentUserManagedEhcache.java:148) [ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at com.terracotta.qa.ehcache.validation.ValidatingGetOperation.exec(ValidatingGetOperation.java:57) [classes/:na]
	at io.rainfall.ScenarioRun.initStatistics(ScenarioRun.java:284) [rainfall-core-1.1.3-20160703.112848-15.jar:na]
	at io.rainfall.ScenarioRun.start(ScenarioRun.java:135) [rainfall-core-1.1.3-20160703.112848-15.jar:na]
	at com.terracotta.qa.ehcache.tiering.test.UserManagedPersistenceCacheTest$2.call(UserManagedPersistenceCacheTest.java:172) [test-classes/:na]
	at java.util.concurrent.FutureTask.run(FutureTask.java:262) [na:1.7.0_79]
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1145) [na:1.7.0_79]
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:615) [na:1.7.0_79]
	at java.lang.Thread.run(Thread.java:745) [na:1.7.0_79]
Caused by: java.lang.NullPointerException: null
	at java.util.concurrent.ConcurrentHashMap.put(ConcurrentHashMap.java:1124) ~[na:1.7.0_79]
	at org.ehcache.impl.serialization.CompactJavaSerializer$OIS.readClassDescriptor(CompactJavaSerializer.java:239) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at java.io.ObjectInputStream.readNonProxyDesc(ObjectInputStream.java:1601) ~[na:1.7.0_79]
	at java.io.ObjectInputStream.readClassDesc(ObjectInputStream.java:1517) ~[na:1.7.0_79]
	at java.io.ObjectInputStream.readOrdinaryObject(ObjectInputStream.java:1771) ~[na:1.7.0_79]
	at java.io.ObjectInputStream.readObject0(ObjectInputStream.java:1350) ~[na:1.7.0_79]
	at java.io.ObjectInputStream.readObject(ObjectInputStream.java:370) ~[na:1.7.0_79]
	at org.ehcache.impl.serialization.CompactJavaSerializer.read(CompactJavaSerializer.java:131) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.impl.internal.store.offheap.LazyOffHeapValueHolder.forceDeserialization(LazyOffHeapValueHolder.java:99) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.impl.internal.store.offheap.LazyOffHeapValueHolder.value(LazyOffHeapValueHolder.java:53) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.impl.internal.store.heap.OnHeapStore.cloneValueHolder(OnHeapStore.java:1455) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.impl.internal.store.heap.OnHeapStore.importValueFromLowerTier(OnHeapStore.java:1445) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	at org.ehcache.impl.internal.store.heap.OnHeapStore.getOrComputeIfAbsent(OnHeapStore.java:703) ~[ehcache-3.1.0-SNAPSHOT.jar:3.1.0-SNAPSHOT a6bf3bda6216c6b190b53114a46fbd3e00709639]
	... 11 common frames omitted
````

--- a/impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java
+++ b/impl/src/main/java/org/ehcache/config/builders/UserManagedCacheBuilder.java
@@ -51,6 +51,7 @@ import org.ehcache.core.spi.LifeCycled;
 import org.ehcache.core.spi.LifeCycledAdapter;
 import org.ehcache.core.internal.service.ServiceLocator;
 import org.ehcache.core.spi.store.Store;
+import org.ehcache.spi.persistence.PersistableResourceService;
 import org.ehcache.spi.service.ServiceProvider;
 import org.ehcache.spi.copy.Copier;
 import org.ehcache.spi.loaderwriter.CacheLoaderWriter;
@@ -201,13 +202,15 @@ public class UserManagedCacheBuilder<K, V, T extends UserManagedCache<K, V>> imp
     CacheConfiguration<K, V> cacheConfig = new BaseCacheConfiguration<K, V>(keyType, valueType, evictionAdvisor,
         classLoader, expiry, resourcePools);
 
+    List<LifeCycled> lifeCycledList = new ArrayList<LifeCycled>();
+
     Set<ResourceType<?>> resources = resourcePools.getResourceTypeSet();
     boolean persistent = resources.contains(DISK);
     if (persistent) {
       if (id == null) {
         throw new IllegalStateException("Persistent user managed caches must have an id set");
       }
-      LocalPersistenceService persistenceService = serviceLocator.getService(LocalPersistenceService.class);
+      final LocalPersistenceService persistenceService = serviceLocator.getService(LocalPersistenceService.class);
       if (!resourcePools.getPoolForResource(ResourceType.Core.DISK).isPersistent()) {
         try {
           persistenceService.destroy(id);
@@ -216,13 +219,18 @@ public class UserManagedCacheBuilder<K, V, T extends UserManagedCache<K, V>> imp
         }
       }
       try {
-        serviceConfigsList.add(persistenceService.getPersistenceSpaceIdentifier(id, cacheConfig));
+        final PersistableResourceService.PersistenceSpaceIdentifier<?> identifier = persistenceService.getPersistenceSpaceIdentifier(id, cacheConfig);
+        lifeCycledList.add(new LifeCycledAdapter() {
+          @Override
+          public void close() throws Exception {
+            persistenceService.releasePersistenceSpaceIdentifier(identifier);
+          }
+        });
+        serviceConfigsList.add(identifier);
       } catch (CachePersistenceException cpex) {
         throw new RuntimeException("Unable to create persistence space for cache " + id, cpex);
       }
     }
-
-    List<LifeCycled> lifeCycledList = new ArrayList<LifeCycled>();
 
     Serializer<K> keySerializer = this.keySerializer;
     Serializer<V> valueSerializer = this.valueSerializer;

--- a/integration-test/src/test/java/org/ehcache/integration/PersistentUserManagedCacheTest.java
+++ b/integration-test/src/test/java/org/ehcache/integration/PersistentUserManagedCacheTest.java
@@ -1,0 +1,86 @@
+/*
+ * Copyright Terracotta, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.ehcache.integration;
+
+import org.ehcache.PersistentUserManagedCache;
+import org.ehcache.config.builders.ResourcePoolsBuilder;
+import org.ehcache.config.builders.UserManagedCacheBuilder;
+import org.ehcache.config.units.EntryUnit;
+import org.ehcache.config.units.MemoryUnit;
+import org.ehcache.core.spi.service.LocalPersistenceService;
+import org.ehcache.impl.config.persistence.DefaultPersistenceConfiguration;
+import org.ehcache.impl.config.persistence.UserManagedPersistenceContext;
+import org.ehcache.impl.persistence.DefaultLocalPersistenceService;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+import java.io.File;
+import java.io.Serializable;
+
+import static org.hamcrest.Matchers.notNullValue;
+import static org.junit.Assert.assertThat;
+
+/**
+ * PersistentUserManagedCacheTest
+ */
+public class PersistentUserManagedCacheTest {
+
+  @Rule
+  public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+  @Test
+  public void recoversWithSerializableType() throws Exception {
+    File folder = temporaryFolder.newFolder("cache-persistence-store");
+    {
+      LocalPersistenceService persistenceService = new DefaultLocalPersistenceService(new DefaultPersistenceConfiguration(folder));
+      PersistentUserManagedCache<Long, Foo> cache = UserManagedCacheBuilder.newUserManagedCacheBuilder(Long.class, Foo.class)
+          .with(new UserManagedPersistenceContext<Long, Foo>("cache-name", persistenceService))
+          .withResourcePools(ResourcePoolsBuilder.newResourcePoolsBuilder()
+              .heap(10L, EntryUnit.ENTRIES)
+              .disk(10L, MemoryUnit.MB, true))
+          .build(true);
+      cache.put(1L, new Foo(1));
+      cache.close();
+      persistenceService.stop();
+    }
+
+    {
+      LocalPersistenceService persistenceService = new DefaultLocalPersistenceService(new DefaultPersistenceConfiguration(folder));
+      PersistentUserManagedCache<Long, Foo> cache = UserManagedCacheBuilder.newUserManagedCacheBuilder(Long.class, Foo.class)
+          .with(new UserManagedPersistenceContext<Long, Foo>("cache-name", persistenceService))
+          .withResourcePools(ResourcePoolsBuilder.newResourcePoolsBuilder()
+              .heap(10L, EntryUnit.ENTRIES)
+              .disk(10L, MemoryUnit.MB, true))
+          .build(true);
+      //It fails here
+      assertThat(cache.get(1L), notNullValue());
+      cache.close();
+      persistenceService.stop();
+      cache.destroy();
+    }
+  }
+
+  private static class Foo implements Serializable {
+
+    private final int i;
+
+    Foo(int i) {
+      this.i = i;
+    }
+  }
+}

